### ISSUE TITLE
Timeouts - puppet master / agent

### ISSUE THREAD

**Issue by robh007** [biemond/biemond-orawls#133]
Just a general question to see if anyone as seen a similar issue.

I appear to be receiving the following timeouts when running puppet in agent / master. This is just an example of the timeouts from the daemon process. Are you waiting for certain output to be return from the WLST process? I've taken a couple of WLST scripts and ran them independent against the admin server and they execute fine.

Have you seen anything similar?

5161625-Mar 20 14:42:09 node-wls puppet-agent[675]: domain found default
5161692-Mar 20 14:42:09 node-wls puppet-agent[675]: Starting the wls daemon for domain default
5161781-Mar 20 14:42:09 node-wls puppet-agent[675]: Connecting to wls on url t3://node-wls:7001
5161892-Mar 20 14:42:09 node-wls puppet-agent[675]: Executing wls-script /tmp/wlstScript20150320-675-u797pe-0.py
5161999-Mar 20 14:42:09 node-wls puppet-agent[675]: Connecting to wls on url t3://node-wls:7001
5162110:Mar 20 14:43:12 node-wls puppet-agent[675]: Could not prefetch wls_user provider 'simple': timeout on reading expected output from daemon process. 

I'm running this on Centos 6.6 I was initially running Puppet version 3.6.2, I've updated to the latest version 3.7.4. I've also updated ORAWLS & Easy_type.

**Comment by biemond** [comment]
Hi,

This is a safety feature of the wlst domain daemon else it would wait for ever.  probably the wlst  script action crashed , this is because  I didn't handle the exception , a certain custom type attribute combination or wlst is in unknown state and the adminserver should be restarted before it can continue .
The first 2 errors can be solved when I know the error and the last one is a configuration error

So try to run puppet in debug and trace mode and you will see the output or look at the weblogic console and see if there are pending changes 

Hope this helps

**Comment by robh007** [comment]
Many thanks for the swift response, I'll have a closer look at the admin console

**Comment by robh007** [comment]
I cant see any issues with the WLS configuration, I get the following when running --debug --trace. I have the same configuration working with a couple of virtual machines, but then I have this issue with physical machines. Both environments have been created in the same way. Any further avenues to explore would be useful. I'm going to try running a local puppet run rather than communicating with the master.

Thanks Rob.

Notice: /Stage[main]/Startwls/Orawls::Control[startWLSAdminServer]/Wls_adminserver[startWLSAdminServer:AdminServer]/ensure: ensure changed 'stop' to 'start'
Debug: /Stage[main]/Startwls/Orawls::Control[startWLSAdminServer]/Wls_adminserver[startWLSAdminServer:AdminServer]: The container Orawls::Control[startWLSAdminServer] will propagate my refresh event
Debug: Orawls::Control[startWLSAdminServer]: The container Class[Startwls] will propagate my refresh event
Debug: Class[Startwls]: The container Stage[main] will propagate my refresh event
Debug: Exec[execwlst Wls12cstoreUserConfig.py](provider=posix): Executing check 'ls -l /home/wls/wls-DevLinux1036Domain-WebLogicConfig.properties'
Debug: Executing 'ls -l /home/wls/wls-DevLinux1036Domain-WebLogicConfig.properties'
Debug: /Stage[main]/Userconfig/Orawls::Storeuserconfig[Wls12c]/Exec[execwlst Wls12cstoreUserConfig.py]/unless: -rw-r--r-- 1 wls dba 227 Mar 23 11:51 /home/wls/wls-DevLinux1036Domain-WebLogicConfig.properties
Debug: Prefetching simple resources for wls_machine
Info: index wls_machine
Debug: Closing connection for https://puppetmaster:8140
Debug: Creating new connection for https://puppetmaster:8140
Debug: Starting connection for https://puppetmaster:8140
Debug: Caching connection for https://puppetmaster:8140
Info: Executing: wlstScript with action index
Info: domain found default
Prepare to run: /tmp/wlstScript20150323-1143-10kv0ut-0.py,wls,default,/opt/oracle/middleware/wlserver_10.3,weblogic,weblogic1,t3://node-wls:7001
vvv===================================================================


def quote(text):
    if text:
        return "\"" + str(text).replace("\"", "\"\"") + "\""
    else:
        return ""

m = ls('/Machines',returnMap='true')

f = open("/tmp/wlstScript.out", "w")
print >>f, "name;machinetype;nmtype;listenaddress;listenport;domain"
for token in m:
    print '___'+token+'___'
    cd('/Machines/'+token)
    type = get('Type')
    cd('NodeManager/'+token)

    listenAddress = get('ListenAddress')
    if not (listenAddress):
       listenAddress = ''

    print >>f, ";".join(map(quote, [domain+'/'+token,type,get('NMType'),listenAddress,str(get('ListenPort')),domain]))

f.close()
print "~~~~COMMAND SUCCESFULL~~~~"


^^^====================================================================
Info: Starting the wls daemon for domain default
Debug: Passing domain default
Debug: Passing credintials to WLST
Info: Connecting to wls on url t3://node-wls:7001
Info: Executing wls-script /tmp/wlstScript20150323-1143-10kv0ut-0.py with timeout = 120
Debug: Passing domain default
Debug: Passing credintials to WLST
Info: Connecting to wls on url t3://node-wls:7001
Debug: CLASSPATH=/opt/oracle/middleware/patch_wls1036/profiles/default/sys_manifest_classpath/weblogic_patch.jar:/opt/oracle/middleware/patch_ocp371/profiles/default/sys_manifest_classpath/weblogic_patch.jar:/usr/java/jdk1.7.0_67/lib/tools.jar:/opt/oracle/middleware/wlserver_10.3/server/lib/weblogic_sp.jar:/opt/oracle/middleware/wlserver_10.3/server/lib/weblogic.jar:/opt/oracle/middleware/modules/features/weblogic.server.modules_10.3.6.0.jar:/opt/oracle/middleware/wlserver_10.3/server/lib/webservices.jar:/opt/oracle/middleware/modules/org.apache.ant_1.7.1/lib/ant-all.jar:/opt/oracle/middleware/modules/net.sf.antcontrib_1.1.0.0_1-0b2/lib/ant-contrib.jar:

Debug:

Debug: PATH=/opt/oracle/middleware/wlserver_10.3/server/bin:/opt/oracle/middleware/modules/org.apache.ant_1.7.1/bin:/usr/java/jdk1.7.0_67/jre/bin:/usr/java/jdk1.7.0_67/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/home/wls/bin

Debug:

Debug: Your environment has been set.

Debug:

Debug: Initializing WebLogic Scripting Tool (WLST) ...

Debug:

Debug: Welcome to WebLogic Server Administration Scripting Shell

Debug:

Debug: Type help() for help on available commands

Debug:

Debug: wls:/offline> wls:/offline> wls:/offline> wls:/offline> Connecting to t3://node-wls:7001 with userid weblogic ...

Debug: Successfully connected to Admin Server 'AdminServer' that belongs to domain 'DevLinux1036Domain'.

Debug:

Debug: Warning: An insecure protocol was used to connect to the

Debug: server. To ensure on-the-wire security, the SSL port or

Debug: Admin port should be used instead.

Debug:

Debug: wls:/DevLinux1036Domain/serverConfig> wls:/DevLinux1036Domain/serverConfig> wls:/DevLinux1036Domain/serverConfig> wls:/DevLinux1036Domain/serverConfig> Connecting to t3://node-wls:7001 with userid weblogic ...

Debug: Successfully connected to Admin Server 'AdminServer' that belongs to domain 'DevLinux1036Domain'.

Debug:

Debug: Warning: An insecure protocol was used to connect to the

Debug: server. To ensure on-the-wire security, the SSL port or

Debug: Admin port should be used instead.

Debug:

Debug: wls:/DevLinux1036Domain/serverConfig> dr--   LocalMachine

Debug:

Error: Could not prefetch wls_machine provider 'simple': timeout on reading expected output from daemon process.
/var/lib/puppet/lib/easy_type/daemon.rb:77:in `timed_readline'
/var/lib/puppet/lib/easy_type/daemon.rb:62:in `sync'
/var/lib/puppet/lib/utils/wls_daemon.rb:36:in `execute_script'
/var/lib/puppet/lib/utils/wls_access.rb:93:in `execute_wlst'
/var/lib/puppet/lib/utils/wls_access.rb:39:in `wlst'
/var/lib/puppet/lib/utils/wls_access.rb:37:in `each'
/var/lib/puppet/lib/utils/wls_access.rb:37:in `wlst'
/var/lib/puppet/lib/puppet/type/wls_machine.rb:23:in `get_raw_resources'
/var/lib/puppet/lib/easy_type/provider.rb:178:in `instances'
/var/lib/puppet/lib/easy_type/provider.rb:193:in `prefetch'
/usr/lib/ruby/site_ruby/1.8/puppet/transaction.rb:307:in `prefetch'
/usr/lib/ruby/site_ruby/1.8/puppet/transaction.rb:197:in `prefetch_if_necessary'
/usr/lib/ruby/site_ruby/1.8/puppet/transaction.rb:96:in `evaluate'
/usr/lib/ruby/site_ruby/1.8/puppet/graph/relationship_graph.rb:116:in `call'
/usr/lib/ruby/site_ruby/1.8/puppet/graph/relationship_graph.rb:116:in `traverse'
/usr/lib/ruby/site_ruby/1.8/puppet/transaction.rb:138:in `evaluate'
/usr/lib/ruby/site_ruby/1.8/puppet/resource/catalog.rb:169:in `apply'
/usr/lib/ruby/site_ruby/1.8/puppet/util/log.rb:149:in `with_destination'
/usr/lib/ruby/site_ruby/1.8/puppet/transaction/report.rb:112:in `as_logging_destination'
/usr/lib/ruby/site_ruby/1.8/puppet/resource/catalog.rb:168:in `apply'
/usr/lib/ruby/site_ruby/1.8/puppet/configurer.rb:120:in `apply_catalog'
/usr/lib/ruby/site_ruby/1.8/puppet/util.rb:161:in `benchmark'
/usr/lib/ruby/1.8/benchmark.rb:308:in `realtime'
/usr/lib/ruby/site_ruby/1.8/puppet/util.rb:160:in `benchmark'
/usr/lib/ruby/site_ruby/1.8/puppet/configurer.rb:119:in `apply_catalog'
/usr/lib/ruby/site_ruby/1.8/puppet/configurer.rb:227:in `run_internal'
/usr/lib/ruby/site_ruby/1.8/puppet/configurer.rb:134:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/context.rb:64:in `override'
/usr/lib/ruby/site_ruby/1.8/puppet.rb:244:in `override'
/usr/lib/ruby/site_ruby/1.8/puppet/configurer.rb:133:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/agent.rb:47:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/agent/locker.rb:20:in `lock'
/usr/lib/ruby/site_ruby/1.8/puppet/agent.rb:47:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/agent.rb:117:in `with_client'
/usr/lib/ruby/site_ruby/1.8/puppet/agent.rb:44:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/agent.rb:82:in `run_in_fork'
/usr/lib/ruby/site_ruby/1.8/puppet/agent.rb:43:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/application.rb:179:in `call'
/usr/lib/ruby/site_ruby/1.8/puppet/application.rb:179:in `controlled_run'
/usr/lib/ruby/site_ruby/1.8/puppet/agent.rb:41:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/application/agent.rb:356:in `onetime'
/usr/lib/ruby/site_ruby/1.8/puppet/application/agent.rb:322:in `run_command'
/usr/lib/ruby/site_ruby/1.8/puppet/application.rb:381:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/application.rb:507:in `plugin_hook'
/usr/lib/ruby/site_ruby/1.8/puppet/application.rb:381:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/util.rb:488:in `exit_on_fail'
/usr/lib/ruby/site_ruby/1.8/puppet/application.rb:381:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/util/command_line.rb:146:in `run'
/usr/lib/ruby/site_ruby/1.8/puppet/util/command_line.rb:92:in `execute'
/usr/bin/puppet:8

**Comment by biemond** [comment]
interesting

I can see the wlst LS step and it found one LocalMachine
but I don't see this  print '___'+token+'___'

So it looks like it cannot write to tmp or have access to wlstScript.out

f = open("/tmp/wlstScript.out", "w")
    print >>f, "name;machinetype;nmtype;listenaddress;listenport;domain"

Hope this helps

    Debug: wls:/DevLinux1036Domain/serverConfig> dr--   LocalMachine

    Debug:

    Error: Could not prefetch wls_machine provider 'simple': timeout on reading expected output from daemon process.


    def quote(text):
        if text:
            return "\"" + str(text).replace("\"", "\"\"") + "\""
        else:
            return ""

    m = ls('/Machines',returnMap='true')

    f = open("/tmp/wlstScript.out", "w")
    print >>f, "name;machinetype;nmtype;listenaddress;listenport;domain"
    for token in m:
        print '___'+token+'___'
        cd('/Machines/'+token)
        type = get('Type')
        cd('NodeManager/'+token)

        listenAddress = get('ListenAddress')
        if not (listenAddress):
           listenAddress = ''

        print >>f, ";".join(map(quote, [domain+'/'+token,type,get('NMType'),listenAddress,str(get('ListenPort')),domain]))

    f.close()
    print "~~~~COMMAND SUCCESFULL~~~~"

**Comment by robh007** [comment]
Thanks for the pointer Edwin. I've sorted the issue, the wlstScript.out file was present in the /tmp/ filesystem but it had initially been created by the user oracle. The hieradata was changed to use user  wls after the initial install and attempted WLS setup. The wls user was unable to carry out an operation on the file. I was getting an error just not visible to the puppet run, which was in fact a file permission issue.
